
Feature Matrix Analysis
=======================

```{r echo=FALSE}
suppressMessages(library(gplots))
suppressMessages(library(xtable))
source(file.path(Sys.getenv("TUMAGG_HOME"),"utils/plot_utils.R")) # for boxplot
library(xtable)
```

```{r echo=FALSE, cache=TRUE}
data.dir <- "~/chipseq/results/20120816"
load(file.path(data.dir,"pe.pvals.RData"))
load(file.path(data.dir,"pe.cor.RData"))
load(file.path(data.dir,"pe.counts.RData"))
load(file.path(data.dir,"fm.nm.RData"))
load(file.path(data.dir,"fm.eid.RData"))
all.nm <- rownames(fm.nm)
all.eid <- rownames(fm.eid)
```

Some Overall Stats
-------------------

GeneIDs(rows): ```r nrow(fm.eid)```  
Have PolII assessment: ```r length(which(!is.na(fm.eid[,"Max PolII Signal"])))```  
Have Poised assessment: ```r length(which(!is.na(fm.eid[,"Poised at T=0"])))```  
Have expression values: ```r length(which(!is.na(fm.eid[,"Induced"])))```  
Have all three: ```r length(which(!is.na(fm.eid[,"Poised at T=0"]) & !is.na(fm.eid[,"Max PolII Signal"]) & !is.na(fm.eid[,"Induced"]))) ```  

RefSeqs(rows): ```r nrow(fm.nm)```  
Have PolII assessment: ```r length(which(!is.na(fm.nm[,"Max PolII Signal"])))```  
Have Poised assessment: ```r length(which(!is.na(fm.nm[,"Poised at T=0"])))```  
Have expression values: ```r length(which(!is.na(fm.nm[,"Induced"])))```  
Have all three: ```r length(which(!is.na(fm.nm[,"Poised at T=0"]) & !is.na(fm.nm[,"Max PolII Signal"]) & !is.na(fm.nm[,"Induced"]))) ```  

Exon expression categories (gene id counts)
```{r echo=FALSE}
table(fm.eid[,"Qualitative Change - Exon"])
```

3' Expression categories (gene id counts)
```{r echo=FALSE}
table(fm.eid[,"Qualitative Change - Three Prime"])
```

Filters on transcripts and genes
---------------------------------

*Filtering poised signal by overall signal strength*. 
We will use two filters, a "stringent" one and a "moderate one". The stringent filter will retain transcripts whose PolII signal exceeds 4 units at some point in time, and the moderate filter will retain transcripts whose Poll signal exceeds 1 at some time point.
```{r cache=TRUE}
s4.nm <- all.nm[which(as.numeric(fm.nm[,"Max PolII Signal"])>4)] ## stringent signal filter
s1.nm <- all.nm[which(as.numeric(fm.nm[,"Max PolII Signal"])>1)] ## moderate signal filter 
```
The stringent filter yields ```r length(s4.nm)``` RefSeqs, the moderate one ```r length(s1.nm)`` RefSeqs.

Expression Array Platform Comparisons
-------------------------------------

On both arrays: ```r length(which(!is.na(fm.eid[,"Constitutive Expression - Three Prime"])&!is.na(fm.eid[,"Constitutive Expression - Exon"])))``` Gene IDs.

Comparison of three prime (rows) and exon (columns) constitutive expression. 1=constitutive as per Gene Expression Differential Tests, e.g "flat" over all timepoints and above threshold. 0=on the array but not in the "1" set.

```{r echo=FALSE}
table(fm.eid[,"Constitutive Expression - Three Prime"],fm.eid[,"Constitutive Expression - Exon"])
s1 <- names(which((fm.eid[,"Constitutive Expression - Three Prime"]==1)&(fm.eid[,"Constitutive Expression - Exon"]==0)))
s2 <- names(which((fm.eid[,"Constitutive Expression - Exon"]==1)&(fm.eid[,"Constitutive Expression - Three Prime"]==0)))
```

Off-diagonal cases: 3-prime constitutive but not exon constitutive: can be subtle difference of not meeting threshold, or genuine disagreement (exon profile sometimes looks like "noisy" version of 3-prime ). Exon constitute but not 3-prime constitutive: e.g. exon profile barely makes threshold for constitutive (< 2-fold). In general "constitutive" definition is based on "not differentially expressed" and *could* be made more stringent if we really wanted to study truly constiutive genes.

Comparison of three prime (rows) and exon (columns) differential expression. 1=differentially expressed, otherwise 0 (but must be on array to be included)
```{r echo=FALSE}
table(fm.eid[,"Differential Expression - Three Prime"],fm.eid[,"Differential Expression - Exon"])
s1 <- names(which((fm.eid[,"Differential Expression - Three Prime"]==1)&(fm.eid[,"Differential Expression - Exon"]==0)))
s2 <- names(which((fm.eid[,"Differential Expression - Exon"]==1)&(fm.eid[,"Differential Expression - Three Prime"]==0)))
```


Off-diagonal elements arise from thresholding differences. 

Comparison of three prime (rows) and exon (columns) cluster assignments for genes on both arrays. 
```{r echo=FALSE}
table(fm.eid[,"Cluster - Three Prime"],fm.eid[,"Cluster - Exon"])
```


For nearly all cases, maximal cluster in row = maximal cluster in column.
Comparison of three prime (rows) and exon (columns) qualitative change for genes on both arrays. 
```{r echo=FALSE}
table(fm.eid[,"Qualitative Change - Three Prime"],fm.eid[,"Qualitative Change - Exon"])
```
Question: Why are many of the exon CIR being labeled as B by 3-prime array analysis? Answer (possibly): Too conservative in choice of 3-prime threshold? Should we aim for thresholds that  imply  similar  sensitivity for   both  arrays?


Comparison of Gene Expression to PolII
---------------------------------------

Comparison of PolII Signal Intensity Cluster (row) with three prime expression cluster (column). 2780 RefSeqs.
```{r echo=FALSE}
table(fm.nm[,"PolII Signal Intensity Cluster"],fm.nm[,"Cluster - Three Prime"])
```
This table contains ```r sum(sum(table(fm.nm[,"PolII Signal Intensity Cluster"],fm.nm[,"Cluster - Three Prime"]))) ``` entries. As mentioned above (signal intensity clustering section) this includes many low-signal PolII profiles.  


Stringent 
```{r echo=FALSE}
tay <- table(fm.nm[s4.nm,"PolII Signal Intensity Cluster"],fm.nm[s4.nm,"Cluster - Three Prime"])
tay
```
(```r sum(tay)``` of the ```r length(s4.nm)``` RefSeqs have a 3-prime clustering assignment)

Moderate 
```{r echo=FALSE}
tay <- table(fm.nm[s1.nm,"PolII Signal Intensity Cluster"],fm.nm[s1.nm,"Cluster - Three Prime"])
tay
```
(```r sum(tay)``` of the ```r length(s1.nm)`` Refseqs have  3-prime  clustering assignment)

Conclude: Strong correspondence between PolII kinetics and gene expression kinetics. Correspondence becomes more pronounced for genes with greater PolII signals. Late expressed ( > 4hr) genes can be reflected in earlier PolII signal. 

Associations with PolII poising
------------------------------

### Compare against PolII signal clusters.

No restriction overall magnitude of PolII intensity

```{r echo=FALSE}
table(fm.nm[,"PolII Signal Intensity Cluster"],fm.nm[,"Poised at T=0"])
```
Moderate
```{r echo=FALSE}
table(fm.nm[s1.nm,"PolII Signal Intensity Cluster"],fm.nm[s1.nm,"Poised at T=0"])
```
Stringent
```{r echo=FALSE}
table(fm.nm[s4.nm,"PolII Signal Intensity Cluster"],fm.nm[s4.nm,"Poised at T=0"])
```

Conclusion: Poising tends toward increasing PollII kinetics. Eg. for the moderate filter, the expected value based on absence of association is
```r round(expectedcount(table(fm.nm[s1.nm,"PolII Signal Intensity Cluster"],fm.nm[s1.nm,"Poised at T=0"]))["Decreasing","1"])```
for Poised and Decreasing, but only 
```r table(fm.nm[s1.nm,"PolII Signal Intensity Cluster"],fm.nm[s1.nm,"Poised at T=0"])["Decreasing","1"]```
are observed. 
The anti-association between poised and Decreasing may in part be due to their definitions.

If we look at induced signals only
```{r echo=FALSE}
polup <- fm.nm[,"PolII Signal Intensity Cluster"]
names(polup) <- all.nm
polup[which(fm.nm[,"PolII Signal Intensity Cluster"]=="Decreasing")] <- NA
polup[which(fm.nm[,"PolII Signal Intensity Cluster"]=="Low Variation")] <- NA
tay <- table(polup,fm.nm[,"Poised at T=0"])
tay
```

The poised genes are not preferentially associated with subsequent expression kinetics.  Slight skewing towards early profile but barely significant, p-value=
```r fisher.test(tay)$p.value``` with an expected count at 1 hr of 
```r round(expectedcount(tay)["Peak at 1hr","1"])``` poised genes.

*Poising compared with qualitative expression change*  
Using combined expression change from both arrays (RefSeqs)

```{r echo=FALSE} 
table(fm.nm[,"Qualitative Change - Arrays Combined"],fm.nm[,"Poised at T=0"])
```

To simplify, remove decreasing and constitutive genes
```{r echo=FALSE}
upornot <- (fm.nm[,"Qualitative Change - Arrays Combined"]=="Below Threshold") | (fm.nm[,"Qualitative Change - Arrays Combined"]=="Induced")
ups <- rownames(fm.nm)[which(upornot)]
tay <- table(fm.nm[ups,"Qualitative Change - Arrays Combined"],fm.nm[ups,"Poised at T=0"])
tay
```
Association test: p <```r fisher.test(tay)$p.value```.  Strong association between gene induction and poising. Expect value for no association is ```r round(expectedcount(tay)["Induced","1"])``` genes. Holds true also if exon and 3-prime arrays are considered alone.

For Gene IDs, the corresponding values are
```{r echo=FALSE}
upornot <- (fm.eid[,"Qualitative Change - Arrays Combined"]=="Below Threshold") | (fm.eid[,"Qualitative Change - Arrays Combined"]=="Induced")
ups <- rownames(fm.eid)[which(upornot)]
tay <- table(fm.eid[ups,"Qualitative Change - Arrays Combined"],fm.eid[ups,"Poised at T=0"])
tay
```
which is a total of ```r sum(sum(tay))``` Gene IDs.
